custom_components/localtuya/coordinator.py

```diff
@@ -363,6 +363,9 @@ async def close(self):
             self._unsub_refresh = None
 
         await self.abort_connect()
+
+        if self.gateway:
+            self.gateway.filter_subdevices()
```

Comment:
This will cause an exception in the line 354 when closing the gateway on LocalTuya shutdown or reload.

Reply:
This would help:
```python
        if self.gateway and not self.gateway.is_closing:
```

Reply:
Also, `close()` is called for each disabled sub-device. So, it's enough to remove only this particular sub-device from  `self.gateway.sub_devices`.

Reply:
> This will cause an exception in the line 354 when closing the gateway on LocalTuya shutdown or reload.

It won't raises an error, filter_subdevices also made to be used in the future if needed.

Reply:
> Why?

Just like it was "RuntimeError: dictionary changed size during iteration" when I was removing absent sub-devices: gateway walks through `sub_devices` to close them, and each sub-device is removed form `sub_devices`.

Frankly, I don't quite understand what this piece
```python
        self.sub_devices = dict(
            filter(lambda dev: not dev[1].is_closing, self.sub_devices.items())
```
is doing: my Python knowledge lacks here. But, I believe if this (in the `close()`)
```python
        if self.gateway:
            self.gateway.filter_subdevices()
```
replace with this:
```python
        if self.gateway and not self.gateway.is_closing:
            self.gateway.sub_devices.pop(self._node_id)
```
will do the same for disabled sub-devices, do it faster than those lines I don't understand, and will not interfere when unloading LocalTuya.

Reply:
It's different. We are not modifying the dict that is for the loop using it. Instead, we are taking that dict and filtering out unwanted sub-devices.

Reply:
I can't run tests now, but, anyway: why do we need to call `filter_subdevices()` not only when a sub-device is disabled, but also during unloading or shutdown? I mean, why not add ` and not self.gateway.is_closing` into the condition?

Reply:
What's the hurts if we do that, Cleaning up sub-devices dictionary even on unloading is normal process for the gateway as well!

Reply:
OK, it's your project and your users 😄

Reply:
I'm really grateful, I may missed something or actually break something and you helping and point me that on theses stuff. I use to make changes that improve integration for me but now that there is other using it I do take time before test/merging. Unless the performance don't slow down and code is readable and makes sense then it's good 😸